### DISCUSSION TITLE
dired-find-alternate-file creates a new buffer when dirvish is enabled?

### DISCUSSION THREAD

**Discussion by moogah** [alexluigit/dirvish#102] [General]
I'm creating this as a discussion instead of an issue because I'm new to emacs and not totally confident that I'm observing an issue with dirvish and not some other behavior that a more experienced emacs user would understand.

steps to reprouce:

1. Open emacs with emacs -Q
2. Run `M-x eval-buffer` on the following code snippet
3. Open dired with `M-x dired`
4. Use `l` or `M-x dired-find-alternate-file` on a directory
5. Open the list of buffers with `C-x C-b`.  There are two dired buffers in the list.

if you skip step 2 above, `C-x C-b` will show only one dired buffer

```
(defvar bootstrap-version)
(let ((bootstrap-file
       (expand-file-name "straight/repos/straight.el/bootstrap.el" user-emacs-directory))
      (bootstrap-version 5))
  (unless (file-exists-p bootstrap-file)
    (with-current-buffer
	(url-retrieve-synchronously
	 "https://raw.githubusercontent.com/raxod502/straight.el/develop/install.el"
	 'silent 'inhibit-cookies)
      (goto-char (point-max))
      (eval-print-last-sexp)))
  (load bootstrap-file nil 'nomessage))

(setq package-enable-at-startup nil)

(straight-use-package 'use-package)

(toggle-debug-on-error)

(use-package dirvish
  :straight t
  :init
  (dirvish-override-dired-mode)
  :bind
  (:map dirvish-mode-map
	("h" . (lambda () (interactive) (find-alternate-file "..")))
	("l" . dired-find-alternate-file)))

(when (string= system-type "darwin")
  (setq dired-use-ls-dired nil))

(put 'dired-find-alternate-file 'disabled nil)
```

**Comment by alexluigit** [comment]
This is actually an expected behavior.  I had it documented [here](https://github.com/alexluigit/dirvish/blob/main/CUSTOMIZING.org#buffer-management)

But since you asked, I guess we need a better explanation so that people won't be confused.

## Dired

By default, Dired creates a new buffer whenever the user enters a new directory. The opened Dired buffers are reused whenever the user tries to access the associated directory of that buffer.  I think it makes sense.  But doing so leaves the user a tons of Dired buffers after some simple navigations in the file system.

That's why Dired added the `dired-find-alternate-file` command and the `dired-kill-when-opening-new-dired-buffer` option (in Emacs 28), they allow the user to keep only one Dired buffer while navigating the different paths in the file
system.  It is of course a good idea, but the implementation is pretty problematic IMO: it achieves the "only one Dired
buffer" goal by killing the current buffer when selecting a new directory.  Say you are in ~/foo, when you go to ~/bar,
the buffer for ~/foo is killed; when do go to ~/baz, the buffer for ~/bar is killed, and so on and so forth.

From my perspective, this "kill-on-the-go" approach is arguably the incorrect way to manage Dired buffers, for it
**breaks** the buffer reuse mechanism of Dired.  Let's say it takes Dired 2s to open the directory ~/foo (it is common
for huge directories or when the directory is accessed over TRAMP), you then go to ~/bar (notice ~/foo is killed by
now), do something, and want to come back to ~/foo, you'll have to wait 2 whole seconds again. Plus all your marks or
subtree folding state are lost.  Besides, if you happen to have the same Dired buffer displayed in different
windows/tabs/frames, they might be killed unexpectedly.  To me, this doesn't make sense at all.

## Dirvish

Dirvish lives in the form of sessions. Every session has its own directory/preview/utility buffer list, dirvish keeps a
linear record of visited path and directories.  It never kill buffers as long as the session is visible, the buffer
cleaning only happens after you quit the session explicitly, namely, open a file or call `dirvish-quit`.  Now we have
the best of both worlds: buffer reuse and a clean buffer list at the same time, no extra configuraion or commands are
needed.  The other benefit is that commands like `dirvish-history-go-forward` and `dirvish-history-go-backward` happens
immediately, it would be very helpful when working with TRAMP connections.

**Comment by moogah** [comment]
🤦 Clearly I didn't read the documentation thoroughly enough.  Thank you for the detailed explanation!

**Reply by ParetoOptimalDev** [reply]
You aren't alone. I almost switched back to to dired from dirvish because of the workflow change until I found `dirvish-reuse-session` being set to true kept things similar.

**Comment by aikrahguzar** [comment]
A tangential question regarding buffer re-use. Is it possible to write a function `dirvish-buffer-current-for-a-session-p` which takes a `BUFFER-OR-NAME` as a argument and returns non nil when it is current for one of the dirvish sessions?

I think buffer reuse as you describe makes total sense but it leads to a proliferation of buffers presented when switching buffers. If `dirvish-buffer-current-for-a-session-p` can be implemented it can then be passed as a predicate so that only currently relevant dirvish-buffers will be presented. My current alternative to this is to filter out all dirvish buffers and rely on `dired-jump` and that is all I need most of the times but on a few occasions it is useful to be able to switch to a dirvish buffer directly.

**Reply by alexluigit** [reply]
I'm not sure if this is what you are looking for:

```emacs-lisp
(defun dirvish-buffer-index-p (buffer-or-name)
  "Return t if BUFFER-OR-NAME is the index of any dirvish session."
  (memq (get-buffer buffer-or-name)
        (remove nil (mapcar #'cdr (dirvish-get-all 'index-dir)))))
```

How would you use it?

**Reply by aikrahguzar** [reply]
> How would you use it?

I still have to experiment but I think using this functions as the predicate for completing read in the buffer switching function should work. In my case that should boil down to using it as the predicate for a source in `consult-buffer` or changing the doom wrapper for it to use `dirvish-buffer-index-p` as the predicate. 

Another way of achieving the same might be to use notion of real buffer in doom and set the dirvish buffer being navigated away from as unreal and navigated into as real (maybe burying and unburying buffers is milder option). The goal is basically to limit the number of dirvish buffers that are available to commands like `switch-buffer`, `next-buffer` and `previous-buffer` to one buffer per dirvish session.

**Reply by alexluigit** [reply]
@aikrahguzar, in the recent updates dirvish managed to display parent directory and directory preview using "virtual" dired buffers, so we don't have irrelevant dired buffers anymore. Do you still feel the need to filter the buffer list? I'm asking because I'm planning to remove the `dirvish-get-all` from core.

**Reply by aikrahguzar** [reply]
> @aikrahguzar, in the recent updates dirvish managed to display parent directory and directory preview using "virtual" dired buffers, so we don't have irrelevant dired buffers anymore. Do you still feel the need to filter the buffer list? I'm asking because I'm planning to remove the `dirvish-get-all` from core.

Yes, nothing seems to have changed for me on that end. I asked that so that I could have a version of `dired-kill-when-opening-new-dired-buffer` where the buffers dirvish is keeping alive for reuse are not presented to me when I try to switch buffer using `SPC .` in doom for example. I now have a version of `dirvish-buffer-index-p` which filters out non-index buffers as well as dirvish-side buffers.

And I actually use private functions and variables to find out which out buffers in the side session too it would be nice if that was possible in a more straightforward manner, since I also have a hook in `dirvish-setup-hook` which turns off `dired-hide-details-mode` but only if the buffer is a side buffer.

I will actually not mind too much if the `dirvish-get-all` gets removed. I think I can get used to using only `dired-jump` and `dirvish-side` for navigating to dirvish/dired which is what I do most of the time anyway and then I can configure buffer switching functions to never present any dired/dirvish buffers.

**Reply by alexluigit** [reply]
> I now have a version of dirvish-buffer-index-p which filters out non-index buffers as well as dirvish-side buffers.

Did you redefine the `consult--source-buffer` to apply the predicate? Just curious.

> I also have a hook in dirvish-setup-hook which turns off dired-hide-details-mode but only if the buffer is a side buffer.

Can you tell me what file attribute(s) makes you want those details always visible? `file-size` and `file-time` (newly added) are the only attrs I care in most of situations, I added them to `dirvish-attributes` and they will hide themselves when the window is too narrow. As a result, I barely turn off `dired-hide-details-mode`, I press `(` only when I want to checkout file mode or link number.

> I will actually not mind too much if the dirvish-get-all gets removed. 

Great. Just in case you are using it in your `dirvish-buffer-index-p`.

**Reply by aikrahguzar** [reply]
> Did you redefine the `consult--source-buffer` to apply the predicate? Just curious.

No, I don't use consult buffer directly but rather `+vertico/switch-workspace-buffer` from doom's vertico module and getting it to not present all dirvish buffers is a bit involved. I though adding a function to `doom-real-buffer-functions` would be enough but that doesn't work because it is used a single time when buffer is displayed (I think, I am not sure when doom flags the buffers as real.).  

I was already using a modified version of the workspace module so I just modified the `+workspace-contains-buffer-p` to filter out non index buffers. `+workspace-contains-buffer-p` is the function that is used as predicate for sources in `+vertico/switch-workspace-buffer`.

 > Can you tell me what file attribute(s) makes you want those details always visible? `file-size` and `file-time` (newly added) are the only attrs I care in most of situations, I added them to `dirvish-attributes` and they will hide themselves when the window is too narrow. As a result, I barely turn off `dired-hide-details-mode`, I press `(` only when I want to checkout file mode or link number.

I occasionally have to look at the user and file modes but mostly file size. But it is also a matter of taste I think, I don't like all that empty space in the buffer. Especially since I have gotten used to just using `dirvish-side` for the quick things, so one the occasions I open a normal dired buffer, I want the details to be there.

> Great. Just in case you are using it in your `dirvish-buffer-index-p`.

I am using it but I think `dirvish-side` + `dired-jump` is more than a good enough replacement.

**Reply by alexluigit** [reply]
> I was already using a modified version of the workspace module so I just modified the `+workspace-contains-buffer-p` to filter out non index buffers. `+workspace-contains-buffer-p` is the function that is used as predicate for sources in `+vertico/switch-workspace-buffer`.

Thanks for the info. Perhaps I'll make dirvish support the `dired-kill-when-opening-new-dired-buffer` option in the future. 

> But it is also a matter of taste I think, I don't like all that empty space in the buffer. 

That's a fair point. Dired with hide-details-mode doesn't look proportional to the window width.
